File: InteractiveBroadcast_iOS/RTCDemo/LiveRoomViewController.m
//
//  LiveRoomViewController.m
//  RTCDemo
//
//  Created by GongYuhua on 2016/9/12.
//  Copyright © 2016年 Agora. All rights reserved.
//

#import "LiveRoomViewController.h"
#import "KeyCenter.h"

@interface LiveRoomViewController () <
    AgoraRtcEngineDelegate,
    UIPopoverPresentationControllerDelegate>
@property (weak, nonatomic) IBOutlet UILabel *roomNameLabel;
@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray *sessionButtons;
@property (weak, nonatomic) IBOutlet UIButton *audioMuteButton;

@property (weak, nonatomic) IBOutlet UIView *videoView0;
@property (weak, nonatomic) IBOutlet UIView *videoView1;
@property (weak, nonatomic) IBOutlet UIView *videoView2;
@property (weak, nonatomic) IBOutlet UIView *videoView3;
@property (weak, nonatomic) IBOutlet UIView *videoView4;
@property (weak, nonatomic) IBOutlet UIView *videoView5;
@property (weak, nonatomic) IBOutlet UIView *videoView6;

@property (strong, nonatomic) AgoraRtcEngineKit *rtcEngine;
@property (assign, nonatomic) BOOL isBroadcaster;
@property (assign, nonatomic) BOOL isMuted;
@property (assign, nonatomic) BOOL shouldStereo;
@property (strong, nonatomic) NSMutableArray<AgoraRtcVideoCanvas *> *videoCanvases;
@property (strong, nonatomic) NSArray<UIView *> *videoViews;
@end

@implementation LiveRoomViewController
- (BOOL)isBroadcaster {
    return self.clientRole == AgoraRtc_ClientRole_Broadcaster;
}

- (void)setClientRole:(AgoraRtcClientRole)clientRole {
    _clientRole = clientRole;
    [self updateButtonsVisiablity];
}

- (void)setIsMuted:(BOOL)isMuted {
    _isMuted = isMuted;
    [self.rtcEngine muteLocalAudioStream:isMuted];
    [self.audioMuteButton setImage:[UIImage imageNamed:(self.isMuted ? @"btn_mute_cancel" : @"btn_mute")] forState:UIControlStateNormal];
}

- (NSArray<AgoraRtcVideoCanvas *> *)displayCanvases {
    NSArray *displayCanvases;
    if (!self.isBroadcaster && self.videoCanvases.count > 0) {
        displayCanvases = [self.videoCanvases subarrayWithRange:NSMakeRange(1, self.videoCanvases.count - 1)];
    } else {
        displayCanvases = self.videoCanvases;
    }
    return displayCanvases;
}

//MARK: - lift cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    self.videoCanvases = [[NSMutableArray alloc] init];
    self.videoViews = @[self.videoView0];
    
    self.roomNameLabel.text = self.roomName;
    [self updateButtonsVisiablity];
    
    [self loadAgoraKit];
}

//MARK: - user action
- (IBAction)doSwitchCameraPressed:(UIButton *)sender {
    [self.rtcEngine switchCamera];
}

- (IBAction)doMutePressed:(UIButton *)sender {
    self.isMuted = !self.isMuted;
}

- (IBAction)doStereoPressed:(UIButton *)sender {
    self.shouldStereo = !self.shouldStereo;
}

- (IBAction)doDoubleTapped:(UITapGestureRecognizer *)sender {
    CGPoint location = [sender locationInView:self.view];
    
    AgoraRtcVideoCanvas *tappedCanvas;
    for (AgoraRtcVideoCanvas *canvas in self.displayCanvases) {
        CGRect rect = [self.view convertRect:canvas.view.frame fromView:canvas.view];
        if (CGRectContainsPoint(rect, location)) {
            tappedCanvas = canvas;
            break;
        }
    }
    
    if (tappedCanvas) {
        
    }
}

- (IBAction)doLeavePressed:(UIButton *)sender {
    [self leaveChannel];
}

//MARK: -
- (void)updateInterface {
    for (AgoraRtcVideoCanvas *canvas in self.videoCanvases) {
        [canvas.view removeFromSuperview];
    }
    
    NSArray *displayCanvases = self.displayCanvases;
    for (NSInteger i = 0; i < displayCanvases.count && i < self.videoViews.count; ++i) {
        UIView *hostingView = [displayCanvases[i] view];
        UIView *videoView = self.videoViews[i];
        hostingView.frame = videoView.bounds;
        [videoView addSubview:hostingView];
    }
    
    [self setStreamTypeForCanvases:displayCanvases];
}

- (void)setStreamTypeForCanvases:(NSArray<AgoraRtcVideoCanvas *> *)canvases {
    if (!canvases.count) {
        return;
    }
    
    AgoraRtcVideoCanvas *firstCanvas = canvases.firstObject;
    [self.rtcEngine setRemoteVideoStream:firstCanvas.uid type:AgoraRtc_VideoStream_High];
    
    for (NSInteger i = 1; i < canvases.count; ++i) {
        AgoraRtcVideoCanvas *canvas = canvases[i];
        [self.rtcEngine setRemoteVideoStream:canvas.uid type:AgoraRtc_VideoStream_Low];
    }
}

- (void)addLocalCanvas {
    AgoraRtcVideoCanvas *localCanvas = [[AgoraRtcVideoCanvas alloc] init];
    localCanvas.uid = 0;
    localCanvas.view = [[UIView alloc] init];
    localCanvas.renderMode = AgoraRtc_Render_Hidden;
    
    [self.videoCanvases addObject:localCanvas];
    [self.rtcEngine setupLocalVideo:localCanvas];
    [self updateInterface];
}

- (AgoraRtcVideoCanvas *)videoCanvasOfUid:(NSUInteger)uid {
    for (AgoraRtcVideoCanvas *canvas in self.videoCanvases) {
        if (canvas.uid == uid) {
            return canvas;
        }
    }
    
    AgoraRtcVideoCanvas *newCanvas = [[AgoraRtcVideoCanvas alloc] init];
    newCanvas.uid = uid;
    newCanvas.view = [[UIView alloc] init];
    newCanvas.renderMode = AgoraRtc_Render_Hidden;
    [self.videoCanvases addObject:newCanvas];
    return newCanvas;
}

- (void)leaveChannel {
    [self setIdleTimerActive:YES];
    
    [self.rtcEngine setupLocalVideo:nil];
    [self.rtcEngine leaveChannel:nil];
    if (self.isBroadcaster) {
        [self.rtcEngine stopPreview];
    }
    
    if ([self.delegate respondsToSelector:@selector(liveVCNeedClose:)]) {
        [self.delegate liveVCNeedClose:self];
    }
}

//MARK: - Agora Media SDK
- (void)loadAgoraKit {
    self.rtcEngine = [AgoraRtcEngineKit sharedEngineWithAppId:[KeyCenter AppId] delegate:self];
    [self.rtcEngine setChannelProfile:AgoraRtc_ChannelProfile_LiveBroadcasting];
    [self.rtcEngine enableDualStreamMode:YES];
    [self.rtcEngine enableVideo];
    [self.rtcEngine setVideoProfile:self.videoProfile swapWidthAndHeight:YES];
    [self.rtcEngine setClientRole:self.clientRole withKey:nil];
    
    if (self.isBroadcaster) {
        [self.rtcEngine startPreview];
    }
    
    [self addLocalCanvas];
    
    int code = [self.rtcEngine joinChannelByKey:nil channelName:self.roomName info:nil uid:0 joinSuccess:nil];
    if (code == 0) {
        [self setIdleTimerActive:NO];
    } else {
        [self alertString:[NSString stringWithFormat:@"Join channel failed: %d", code]];
    }
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine firstRemoteVideoDecodedOfUid:(NSUInteger)uid size:(CGSize)size elapsed:(NSInteger)elapsed {
    AgoraRtcVideoCanvas *userCanvas = [self videoCanvasOfUid:uid];
    [self.rtcEngine setupRemoteVideo:userCanvas];
    
    [self updateInterface];
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine firstLocalVideoFrameWithSize:(CGSize)size elapsed:(NSInteger)elapsed {
    [self updateInterface];
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine didOfflineOfUid:(NSUInteger)uid reason:(AgoraRtcUserOfflineReason)reason {
    AgoraRtcVideoCanvas *deleteCanvas;
    for (AgoraRtcVideoCanvas *canvas in self.videoCanvases) {
        if (canvas.uid == uid) {
            deleteCanvas = canvas;
        }
    }
    
    if (deleteCanvas) {
        [self.videoCanvases removeObject:deleteCanvas];
        [deleteCanvas.view removeFromSuperview];
        [self updateInterface];
    }
}

//MARK: -
- (void)updateButtonsVisiablity {
    for (UIButton *button in self.sessionButtons) {
        button.hidden = !self.isBroadcaster;
    }
}

- (void)setIdleTimerActive:(BOOL)active {
    [UIApplication sharedApplication].idleTimerDisabled = !active;
}

- (void)alertString:(NSString *)string {
    if (!string.length) {
        return;
    }
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil message:string preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

//MARK: - popover delegate
- (UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller {
    return UIModalPresentationNone;
}
@end

File: AgoraFaceU_iOS/OpenLive/LiveRoomViewController.swift
//
//  LiveRoomViewController.swift
//  OpenLive
//
//  Created by GongYuhua on 6/25/16.
//  Copyright © 2016 Agora. All rights reserved.
//

import UIKit

protocol LiveRoomVCDelegate: NSObjectProtocol {
    func liveVCNeedClose(_ liveVC: LiveRoomViewController)
}

class LiveRoomViewController: UIViewController {
    
    @IBOutlet weak var roomNameLabel: UILabel!
    @IBOutlet weak var remoteContainerView: UIView!
    @IBOutlet weak var broadcastButton: UIButton!

    @IBOutlet weak var audioMuteButton: UIButton!
    @IBOutlet var sessionButtons: [UIButton]!
    
    @IBOutlet weak var faceUContainerViewBottomConstraint: NSLayoutConstraint!
    @IBOutlet weak var faceUCollectView: UICollectionView!
    @IBOutlet weak var faceUVIewContainer: UIView!
    var roomName: String!
    var clientRole = AgoraRtcClientRole.clientRole_Audience {
        didSet {
            updateButtonsVisiablity()
        }
    }
    var videoProfile: AgoraRtcVideoProfile!
    weak var delegate: LiveRoomVCDelegate?
    
    //MARK: - engine & session view
    var rtcEngine: AgoraRtcEngineKit!
    var agoraEnhancer: AgoraYuvEnhancerObjc?
    var videoSrc: AgoraVideoSource?

    fileprivate var isBroadcaster: Bool {
        return clientRole == .clientRole_Broadcaster
    }
    fileprivate var isMuted = false {
        didSet {
            rtcEngine?.muteLocalAudioStream(isMuted)
            audioMuteButton?.setImage(UIImage(named: isMuted ? "btn_mute_cancel" : "btn_mute"), for: UIControlState())
        }
    }
    
    fileprivate var videoSessions = [VideoSession]() {
        didSet {
            guard remoteContainerView != nil else {
                return
            }
            updateInterfaceWithAnimation(true)
        }
    }
    fileprivate var fullSession: VideoSession? {
        didSet {
            if fullSession != oldValue && remoteContainerView != nil {
                updateInterfaceWithAnimation(true)
            }
        }
    }
    
    fileprivate let viewLayouter = VideoViewLayouter()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        roomNameLabel.text = roomName
        updateButtonsVisiablity()
        
        loadAgoraKit()
        faceUCollectView.delegate = self
        faceUCollectView.dataSource = self
        initLayout()
        faceUContainerViewBottomConstraint.constant = -faceUVIewContainer.frame.height
        
    }
    
    //MARK: - user action
    
    @IBAction func pressFaceUButton(_ sender: UIButton) {
        showFaceUView()
    }
    
    @IBAction func pressFaceUCloseButton(_ sender: UIButton) {
        hideFaceUView()
        FaceU.sharedInstance.closeFaceEffects()
    }
    @IBAction func doSwitchCameraPressed(_ sender: UIButton) {
//        rtcEngine?.switchCamera()
        FaceU.sharedInstance.switchCamera()
    }
    
    @IBAction func doMutePressed(_ sender: UIButton) {
        isMuted = !isMuted
    }
    
    @IBAction func doBroadcastPressed(_ sender: UIButton) {
        if isBroadcaster {
            clientRole = .clientRole_Audience
            videoSrc?.detach()
            FaceU.sharedInstance.stopCapture()

        } else {
            clientRole = .clientRole_Broadcaster
            FaceU.sharedInstance.startCapture()
            videoSrc?.attach()
        }
        rtcEngine.setClientRole(clientRole, withKey: nil)
        updateInterfaceWithAnimation(true)
    }
    
    @IBAction func doDoubleTapped(_ sender: UITapGestureRecognizer) {
        if fullSession == nil {
            //将双击到的session全屏
            if let tappedSession = viewLayouter.reponseSessionOfGesture(sender, inSessions: videoSessions, inContainerView: remoteContainerView) {
                fullSession = tappedSession
            }
        } else {
            fullSession = nil
        }
    }
    
    @IBAction func doLeavePressed(_ sender: UIButton) {
        leaveChannel()
    }
    
    var preLocalSessionFrame: CGRect?
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        guard let localSession = self.fetchSessionOfUid(0) else {
            return
        }
        if preLocalSessionFrame != nil && localSession.canvas.view.frame != preLocalSessionFrame {
            print("frame change to \(localSession.canvas.view.frame)")
            FaceU.sharedInstance.removeFaceUFromLocalSession()
            FaceU.sharedInstance.addFaceUToLocalSession(localView: localSession.canvas.view)
        }
        preLocalSessionFrame = localSession.canvas.view.frame

        
    }
}

private extension LiveRoomViewController {
    
    func updateButtonsVisiablity() {
        guard sessionButtons != nil else {
            return
        }
        broadcastButton?.setImage(UIImage(named: isBroadcaster ? "btn_join_cancel" : "btn_join"), for: UIControlState())
        
        for button in sessionButtons {
            button.isHidden = !isBroadcaster
        }
    }
    
    func leaveChannel() {
        setIdleTimerActive(true)
        
//        rtcEngine.setupLocalVideo(nil)
        
        rtcEngine.leaveChannel(nil)
        if isBroadcaster {
            videoSrc?.detach()
            FaceU.sharedInstance.stopCapture()
//            rtcEngine.stopPreview()
        }
        
        for session in videoSessions {
            session.hostingView.removeFromSuperview()
        }
        videoSessions.removeAll()
        
        agoraEnhancer?.turnOff()
        
        delegate?.liveVCNeedClose(self)
    }
    
    func setIdleTimerActive(_ active: Bool) {
        UIApplication.shared.isIdleTimerDisabled = !active
    }
    
    func alertString(_ string: String) {
        guard !string.isEmpty else {
            return
        }
        
        let alert = UIAlertController(title: nil, message: string, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: nil))
        present(alert, animated: true, completion: nil)
    }

}

private extension LiveRoomViewController {
    func updateInterfaceWithAnimation(_ animation: Bool) {
        if animation {
            
            UIView.animate(withDuration: 0.3, animations: { [weak self] _ in
                self?.updateInterface()
                self?.remoteContainerView.layoutIfNeeded()

                })
        } else {
            updateInterface()
        }
    }
    
    
    
    func updateInterface() {
        var displaySessions = videoSessions
        if !isBroadcaster && !displaySessions.isEmpty {
            displaySessions.removeFirst()
        }
        viewLayouter.layoutSessions(displaySessions, fullSession: fullSession, inContainer: remoteContainerView)
        setStreamTypeForSessions(displaySessions, fullSession: fullSession)

    }
    
    func setStreamTypeForSessions(_ sessions: [VideoSession], fullSession: VideoSession?) {
        if let fullSession = fullSession {
            for session in sessions {
                rtcEngine.setRemoteVideoStream(UInt(session.uid), type: (session == fullSession ? .videoStream_High : .videoStream_Low))
            }
        } else {
            for session in sessions {
                rtcEngine.setRemoteVideoStream(UInt(session.uid), type: .videoStream_High)
            }
        }
    }
    func addLocalSession() {
        let localSession = VideoSession.localSession()
        videoSessions.append(localSession)
//        rtcEngine.setupLocalVideo(localSession.canvas)
        FaceU.sharedInstance.addFaceUToLocalSession(localView: localSession.canvas.view)
    }
    
    func fetchSessionOfUid(_ uid: Int64) -> VideoSession? {
        for session in videoSessions {
            if session.uid == uid {
                return session
            }
        }
        
        return nil
    }
    
    func videoSessionOfUid(_ uid: Int64) -> VideoSession {
        if let fetchedSession = fetchSessionOfUid(uid) {
            return fetchedSession
        } else {
            let newSession = VideoSession(uid: uid)
            videoSessions.append(newSession)
            return newSession
        }
    }
}

//MARK: - Agora Media SDK
private extension LiveRoomViewController {
    
    func loadAgoraKit() {
        
        rtcEngine = AgoraRtcEngineKit.sharedEngine(withAppId: VendorKey, delegate: self)
        rtcEngine.setChannelProfile(.channelProfile_LiveBroadcasting)
        rtcEngine.enableDualStreamMode(true)
        rtcEngine.enableVideo()
        rtcEngine.setVideoProfile(videoProfile, swapWidthAndHeight: true)
        rtcEngine.setClientRole(clientRole, withKey: nil)

        videoSrc = AgoraVideoSource()
        // Init faceU
        if videoSrc != nil {
            FaceU.sharedInstance.configure(videoSource: videoSrc!)
        }
        
        if isBroadcaster {
//            rtcEngine.startPreview()
            videoSrc?.attach()
            FaceU.sharedInstance.startCapture()
        }

        addLocalSession()
        
        let code = rtcEngine.joinChannel(byKey: nil, channelName: roomName, info: nil, uid: 0, joinSuccess: nil)
        if code == 0 {
            setIdleTimerActive(false)
            rtcEngine.setEnableSpeakerphone(true)
        } else {
            DispatchQueue.main.async(execute: {
                self.alertString("Join channel failed: \(code)")
            })
        }
        
        if isBroadcaster {
            let enhancer = AgoraYuvEnhancerObjc()
            enhancer.turnOn()
            self.agoraEnhancer = enhancer
        }
    }
}

extension LiveRoomViewController: AgoraRtcEngineDelegate {
    func rtcEngine(_ engine: AgoraRtcEngineKit!, firstRemoteVideoDecodedOfUid uid: UInt, size: CGSize, elapsed: Int) {
        // display remote view
        let userSession = videoSessionOfUid(Int64(uid))
        rtcEngine.setupRemoteVideo(userSession.canvas)
    }
    
    func rtcEngine(_ engine: AgoraRtcEngineKit!, firstLocalVideoFrameWith size: CGSize, elapsed: Int) {
        if let _ = videoSessions.first {
            updateInterfaceWithAnimation(false)
        }
    }
    
    func rtcEngine(_ engine: AgoraRtcEngineKit!, didOfflineOfUid uid: UInt, reason: AgoraRtcUserOfflineReason) {
        var indexToDelete: Int?
        for (index, session) in videoSessions.enumerated() {
            if session.uid == Int64(uid) {
                indexToDelete = index
            }
        }
        
        if let indexToDelete = indexToDelete {
            let deletedSession = videoSessions.remove(at: indexToDelete)
            deletedSession.hostingView.removeFromSuperview()
            
            if deletedSession == fullSession {
                fullSession = nil
            }
        }
        
        if let _ = videoSessions.first {
            updateInterfaceWithAnimation(false)
        }
    }
}

class FaceUCollectionViewCell: UICollectionViewCell {
    
    @IBOutlet weak var nameLabel: UILabel!
}

extension LiveRoomViewController: UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    
    func initLayout() {
        // Collection view
        let flowLayout = UICollectionViewFlowLayout()
        flowLayout.scrollDirection = UICollectionViewScrollDirection.vertical
        flowLayout.itemSize = CGSize(width: 100, height: 100) // the size will be redefined in UICollectionViewDelegate
        
        // Space between cells and between all cell and view bounds
        flowLayout.minimumInteritemSpacing = 0
        flowLayout.minimumLineSpacing = 0
        self.faceUCollectView.contentInset = UIEdgeInsets.zero
        self.faceUCollectView.collectionViewLayout = flowLayout
    }
    
    
    
    // MARK: UICollectionViewDataSource, UICollectionViewDelegate
    
    /**
     Get number of cells
     
     - parameter collectionView: the collectionView
     - parameter section:        the section index
     
     - returns: the number of image
     */
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return FaceU.sharedInstance.effects.count
    }
    
    
    
    /**
     Get cell
     
     - parameter collectionView:       the collectionView
     - parameter indexPath:            the indexPath
     
     - returns: the cell
     */
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "FaceUCollectionViewCell",
                                                      for: indexPath) as! FaceUCollectionViewCell

        cell.nameLabel.text = FaceU.sharedInstance.effects[indexPath.row].displayName
        return cell
    }
    
    
    
    /**
     Get cell size. Calculates depending on screen size.
     
     - parameter collectionView:       the collectionView
     - parameter collectionViewLayout: the layout
     - parameter indexPath:            the indexPath
     
     - returns: the size
     */
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        
        return collectCellSize()
    }
    
    func collectCellSize() -> CGSize {
        let screenWidth = UIScreen.main.bounds.width
        let n: CGFloat = 4
        
        let width: CGFloat = (screenWidth) / n
        let height: CGFloat = 45
        
        return CGSize(width: width, height: height)
    }
    
    /**
     Show Details screen when cell tapped
     
     - parameter collectionView:       the collectionView
     - parameter indexPath:            the indexPath
     */
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        FaceU.sharedInstance.onFaceEffect(forIndex: indexPath.row)
        hideFaceUView()
        collectionView.deselectItem(at: indexPath, animated: true)
    }

}

//MARK: - faceU view
extension LiveRoomViewController {
    
    func showFaceUView() {
        
        
        guard faceUContainerViewBottomConstraint.constant != 0 else {
            return
        }
        
        
        faceUContainerViewBottomConstraint.constant = 0
        UIView.animate(withDuration: 0.3, animations: {
            self.faceUVIewContainer.alpha = 1
            self.view.layoutIfNeeded()
        })
        
    }
    
    func hideFaceUView() {
        
        guard faceUContainerViewBottomConstraint.constant == 0 else {
            return
        }
        
        faceUContainerViewBottomConstraint.constant = -faceUVIewContainer.frame.height
        
        UIView.animate(withDuration: 0.3, animations: {
            self.faceUVIewContainer.alpha = 0
            self.view.layoutIfNeeded()
        })
    }

}
